internal/blocktx/processor.go

```diff
@@ -662,55 +720,221 @@ func (p *Processor) markTransactionsAsMined(ctx context.Context, blockID uint64,
+	return nil
+}
+
+func (p *Processor) handleStaleBlock(ctx context.Context, block *blocktx_api.Block) (txsToPublish []store.TransactionBlock, err error) {
+	ctx, span := tracing.StartTracing(ctx, "handleStaleBlock", p.tracingEnabled, p.tracingAttributes...)
+	defer func() {
+		tracing.EndTracing(span, err)
+	}()
+
+	staleBlocks, err := p.store.GetStaleChainBackFromHash(ctx, block.Hash)
+	if err != nil {
+		p.logger.Error("unable to get STALE blocks to verify chainwork", slog.String("hash", getHashStringNoErr(block.Hash)), slog.Uint64("height", block.Height), slog.String("err", err.Error()))
+		return nil, err
+	}
+
+	lowestHeight := block.Height
+	if len(staleBlocks) > 0 {
+		lowestHeight = staleBlocks[0].Height
+	}
+
+	longestBlocks, err := p.store.GetLongestChainFromHeight(ctx, lowestHeight)
+	if err != nil {
+		p.logger.Error("unable to get LONGEST blocks to verify chainwork", slog.String("hash", getHashStringNoErr(block.Hash)), slog.Uint64("height", block.Height), slog.String("err", err.Error()))
+		return nil, err
+	}
+
+	staleChainwork := sumChainwork(staleBlocks)
+	longestChainwork := sumChainwork(longestBlocks)
+
+	if longestChainwork.Cmp(staleChainwork) < 0 {
+		p.logger.Info("chain reorg detected", slog.String("hash", getHashStringNoErr(block.Hash)), slog.Uint64("height", block.Height))
+
+		txsToPublish, err = p.performReorg(ctx, staleBlocks, longestBlocks)
+		if err != nil {
+			p.logger.Error("unable to perform reorg", slog.String("hash", getHashStringNoErr(block.Hash)), slog.Uint64("height", block.Height), slog.String("err", err.Error()))
+			return nil, err
 		}
-		err = p.mqClient.PublishMarshal(ctx, MinedTxsTopic, txBlock)
+		return txsToPublish, nil
+	}
+
+	return nil, nil
+}
+
+func (p *Processor) performReorg(ctx context.Context, staleBlocks []*blocktx_api.Block, longestBlocks []*blocktx_api.Block) (txsToPublish []store.TransactionBlock, err error) {
+	ctx, span := tracing.StartTracing(ctx, "performReorg", p.tracingEnabled, p.tracingAttributes...)
+	defer func() {
+		tracing.EndTracing(span, err)
+	}()
+
+	staleHashes := make([][]byte, len(staleBlocks))
+	longestHashes := make([][]byte, len(longestBlocks))
+
+	blockStatusUpdates := make([]store.BlockStatusUpdate, len(longestBlocks)+len(staleBlocks))
+
+	for i, b := range longestBlocks {
+		longestHashes[i] = b.Hash
+
+		b.Status = blocktx_api.Status_STALE
+		update := store.BlockStatusUpdate{Hash: b.Hash, Status: b.Status}
+		blockStatusUpdates[i] = update
+	}
+
+	for i, b := range staleBlocks {
+		staleHashes[i] = b.Hash
+
+		b.Status = blocktx_api.Status_LONGEST
+		update := store.BlockStatusUpdate{Hash: b.Hash, Status: b.Status}
+		blockStatusUpdates[i+len(longestBlocks)] = update
+	}
+
+	err = p.store.UpdateBlocksStatuses(ctx, blockStatusUpdates)
+	if err != nil {
+		return nil, err
+	}
+
+	p.logger.Info("reorg performed successfully")
+
+	registeredTxs, err := p.store.GetRegisteredTxsByBlockHashes(ctx, append(staleHashes, longestHashes...))
```

Comment:
With two requests you will simplfy logic

Reply:
Done